### PR TITLE
Data race fixes

### PR DESCRIPTION
**Pull request by vaskoz**
Fixes for race conditions that exist when GOMAXPROCS>1

To see these errors locally do the following:
```
export GOMAXPROCS=8
go test -race -v ./...
```

I'm running: go version go1.4.2 darwin/amd64

### CODE REVIEW

File: raft/log_test.go [hunk 1]
```diff
@@ -281,9 +281,7 @@ func TestLog_WriteEntriesTo_Snapshot(t *testing.T) {
 	}()
 
 	// Wait for buffer to be written to.
-	for buf.Len() == 0 {
-		runtime.Gosched()
-	}
+	time.Sleep(1 * time.Second)
```
Comment (otoolep) [inline raft/log_test.go]: Why make this change? `Len()` has been locked, so there is no problem spinning on it, right?

File: .gitignore [hunk 2]
```diff
@@ -17,6 +17,7 @@ TAGS
 
 # ignore generated files.
 cmd/influxd/version.go
+cmd/influxd/term
```
Comment (otoolep) [inline .gitignore]: I don't recognise this filename. How was it generated?

Comment (vaskoz) [inline .gitignore]: @otoolep In hindsight, that file shouldn't be gitignored. I believe it was leftover from a test that didn't clean up correctly for me. I will remove this commit.

File: raft/log_test.go (hunk 1 above)
Comment (vaskoz) [inline raft/log_test.go]: I noticed intermittently incorrect results. For instance, I just noticed another with the same pattern in the same file:
```
--- FAIL: TestLog_WriteEntriesTo-8 (0.02s)
	log_test.go:267: unexpected entry count: 1
```
And I'm not sure I understand what you mean by "Len() has been locked". But what seems to be happening is that not all of the data is being written to 'buf' before it gets processed.

**Comment by vaskoz** [comment]
@otoolep I've added another commit, and I've run the tests 5 times with always correct results. Before these changes, the test results and failures were inconsistent: data races caught by the detector and others not caught which just resulted in failing test assertions.
https://gist.github.com/vaskoz/4d46a816ab5adca5645a

File: raft/log_test.go (hunk 1 above)
Comment (otoolep) [inline raft/log_test.go]: I'd need to check the code, but I thought we put a lock into `Len`, such that access to the underlying buffer was atomic (I need to check). And you may have found a different issue too.

Comment (otoolep) [inline raft/log_test.go]: Thanks @vaskoz 

OK, I see what you've done by adding a lock to `Read` as well i.e. `Read` has been locked. With that lock in place, do we still need to replace `runtime.Gosched`? If so, why? 

This is an important change, as it may fix some issues I also have seen in unit testing, and I'm keen to fix it. That's why I want us both to understand this change.

File: raft/log_test.go [hunk 3]
```diff
@@ -401,10 +397,10 @@ func TestCluster_Apply(t *testing.T) {
 	if n := len(c.Logs[0].FSM.(*FSM).Commands); n != 1 {
 		t.Fatalf("unexpected command count(0): %d", n)
 	}
-	if n := len(c.Logs[1].FSM.(*FSM).Commands); n != 1 {
+	if n := len(c.Logs[1].FSM.(*FSM).Commands); n != 0 {
```
Comment (otoolep) [inline raft/log_test.go]: Are these changes deliberate?

**Comment by toddboom** [comment]
The clustering is changing for the next release, so we're closing this out since it won't be applicable any longer.
